hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheService.java

```diff
@@ -228,7 +227,7 @@ public DistributedObject createDistributedObject(String cacheNameWithPrefix) {
                     cacheConfig.setManagerPrefix(HazelcastCacheManager.CACHE_MANAGER_PREFIX);
                 }
 
-                CacheMergePolicyProvider mergePolicyProvider = splitBrainHandlerService.getMergePolicyProvider();
+                CacheMergePolicyProvider mergePolicyProvider = new CacheMergePolicyProvider(nodeEngine);
```

Comment:
We should have a single `CacheMergePolicyProvider` instance per HZ node, which is used for the config check and the split-brain merging. The provider class caches the created merge policy instances, so with a single instance we know for sure that the merge policy instance, which survived the config check is used in the split-brain process later. It also saves us a bunch of `Class.for()` calls and garbage.

That's why the legacy merge policy providers and all other data structures use the single instance of `SplitBrainMergeProvider` from the `NodeEngine`.

Same applies for `IMap` and `ReplicatedMap`.

Reply:
So basically the services or map service context should store the created merge policy provider and make it accessible to the merge process.

Reply:
Maybe we can define a `LegacySplitBrainMergePolicyProvider` interface with a `Object getMergePolicy(String mergePolicyName)` method. Then we can store a single instance in the `AbstractSplitBrainHandlerService` and make it easily accessible for the config checks and merge runnables.

Reply:
changed it to one-instance-per-service.